src/Rules/Keywords/RequireFileExistsRule.php

```diff
@@ -0,0 +1,91 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Keywords;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\Include_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\IdentifierRuleError;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\ShouldNotHappenException;
+use function sprintf;
+use function str_replace;
+use function stream_resolve_include_path;
+
+/**
+ * @implements Rule<Include_>
+ */
+final class RequireFileExistsRule implements Rule
+{
+
+	public function getNodeType(): string
+	{
+		return Include_::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		$errors = [];
+		$paths = $this->resolveFilePaths($node, $scope);
+
+		foreach ($paths as $path) {
+			if (stream_resolve_include_path($path) !== false) {
```

Comment:
This is a clever shortcut but I worry it doesn't work.

Example:

* We're analysing file /var/www/A.php
* We're including 'B.php'
* This file is next to A.php
* include() works but stream_resolve_include_path called from inside PHPStan rule will fail because its PHP code is elsewhere on the disk.
* The result is a false positive.

Can you verify this assumption here? Thanks.

Reply:
You are right. This is actually the reason why I didn't touch the relative paths at all at the beginning and I was showing through the tests that they do not work. The `stream_resolve_include_path` should work exactly like the `include` but the problem is this line from the php documentation: 
> include will finally check in the calling script's own directory and the current working directory before failing.

Unfortunately, we cannot know the actual working directory of application. The problem becomes even worse if you think that different files may have different working directories. 
One solution is for users to call phpstan from the working directory that they want to test against. Another thing that we could do is to make the rule accept an array of working directories that we would check in addition of the `stream_resolve_include_path`.

Since there are a few cases that produce both false positives and false negatives we can disallow the usage of relative paths. A simple `stream_resolve_include_path($path) === $path` should show if the path was relative or not.

Reply:
You can test that it doesn't matter where the PHPStan's code is but instead where what the working directory is by adding in the `require-file-relative-path.php` this line `include 'preload.php'`. This file is in the root directory and depending from which directory you run the phpunit you will have different results.

Reply:
@ondrejmirtes Kind reminder.

Reply:
I'd be fine with getting the same CWD PHPStan has, it's reasonable to assume it's going to be the same for the application.

But your code is missing one use case - "calling script's own directory". That's why you can't use `stream_resolve_include_path` but instead have to check all the same places PHP checks with own code.

Reply:
> But your code is missing one use case - "calling script's own directory".
But can we know the called script? To make it more specific, if my application is being called through the `public/index.php` can the PHPStan understand this?
Or what you are saying is that `stream_resolve_include_path` would also check the calling script's own directory and since in our case the calling script either PHPStan or PHPUnit it might generate false results?

Reply:
I just tested that. It's not a directory with the application entrypoint like index.php, but it's the same directory where the script with the `require` call is. Which means that when we're analysing a `require` AST node, we need to look at `$scope–>getFile()`, get its directory and look for the file in that directory.

Example:

- a.php
- sub/b.php
- sub/c.php
- c.php

a.php includes `sub/b.php` (with absolute path). `sub/b.php` includes `c.php` with a relative path.

`sub/c.php` says `echo "in subdirectory"`.

When we execute `a.php`, the question is: will `c.php` or `sub/c.php` be executed? The answer is `sub/c.php`.

Reply:
You are absolutely correct. At first, I got different results than yours but only because my working directory was in the same level as `a.php` and `c.php`. This made me recheck the priority that file has and here are the results:
1. Working directory has the highest priority `getcwd()`.
2. Include paths are checked second `get_include_path()`.
3. Last but not least, is what you observed. Basically, the folder of the current file is checked.

Reply:
Aaaand i think we are finally close to the end. Fixed in https://github.com/phpstan/phpstan-src/pull/3294/commits/0a53397f3bb9d708f62a25a33814bf84fc1e3080